[minor change] Add support for annotation in aci_rest module (#437)

samiib opened the pull request:
fixes #437

akinross commented on plugins/modules/aci_rest.py:
```diff
@@ -303,6 +309,28 @@ def update_qsl(url, params):
         return url + "?" + "&".join(["%s=%s" % (k, v) for k, v in params.items()])
 
 
+def add_annotation(annotation, payload):
+    """Add annotation to payload only if it has not already been added"""
+    if annotation:
+        for key, val in payload.items():
+            if key in ANNOTATION_UNSUPPORTED:
+                return
+            att = val.get("attributes", {})
+            if "annotation" not in att.keys():
+                att["annotation"] = annotation
+
+
+def add_annotation_xml(annotation, tree):
+    """Add annotation to payload xml only if it has not already been added"""
+    if annotation:
+        for element in tree.iter():
+            if element.tag in ANNOTATION_UNSUPPORTED:
+                return
+            ann = element.get("annotation")
+            if ann is None:
+                element.set("annotation", annotation)
```
Should we write tests for this functionality?

samiib replied:
Good point, yeah I think we should.

samiib replied:
Done! and they pass 😸

akinross commented on plugins/modules/aci_rest.py:
```diff
@@ -284,9 +288,11 @@
+from ansible_collections.cisco.aci.plugins.module_utils.aci import ACIModule, aci_argument_spec, aci_annotation_spec
 from ansible.module_utils._text import to_text
 
+ANNOTATION_UNSUPPORTED = ["tagTag"]
```
I think this list should be expanded with all configurable classes that do not have annotation property, and placed in the constants file

samiib replied:
This is a good idea. However, after we did some analysis we found 830 classes that don't use annotation properties.
Our options are:
1. We could have a big list in the constants file. (Fastest)
2. Include a meta data file and read annotation supported classes from there. (Slower)
3. Have a retry mechanism that checks for errors relating to annotation (More calls to Apic, possibly slowest)
I think 1 is fine but it will be a big list.
@lhercot any thoughts?

lhercot replied:
We can do 1. with a small script that runs as part of the release_pr process to update from the published documentation so we never have to update it manually? In that case might be easier to have a separate file for just that variable but not sure.

samiib replied:
Okay that's a great idea. I'll have to figure out how to implement it, so for now I'll add the scripted release_pr process to the backlog and do option one temporarily.

samiib replied:
Temporary list of classes implemented and scripted process added to backlog.

akinross commented on plugins/modules/aci_rest.py:
```diff
@@ -303,6 +309,28 @@ def update_qsl(url, params):
         return url + "?" + "&".join(["%s=%s" % (k, v) for k, v in params.items()])
 
 
+def add_annotation(annotation, payload):
```
Are we setting annotation in child objects in the payload?

samiib replied:
I thought about it but didn't add support for children. It wasn't specifically asked for but I can.
To be fully featured it would need to recursively add annotation to all children that support it.
Alternatively I could just add it to the first child but not sure how useful that would be.

samiib replied:
@lhercot  - How do you feel about fully implementing annotation to all children that support it?

lhercot replied:
We probably should but that might be complicated.

samiib replied:
@lhercot would you like me to implement it in this PR or add it to the backlog and do a separate PR?

samiib replied:
Turns out it wasn't as complex as I thought so I didn't do a separate PR. Now all children get annotation unless they already have annotation configured. I also added integration testing to verify annotation on child objects.

akinross reviewed (dismissed):
LGTM

sajagana reviewed (changes requested):
cm_add_tenant.proposed.fvTenant.attributes.annotation == "orchestrator:ansible" check is failing.

samiib commented:
> cm_add_tenant.proposed.fvTenant.attributes.annotation == "orchestrator:ansible" check is failing.

@sajagana This check in the xml_string tests should be fixed now.

akinross reviewed (dismissed):
LGTM

sajagana commented on tests/integration/targets/aci_rest/tasks/xml_file.yml:
```diff
@@ -50,6 +50,7 @@
     that:
       - cm_add_tenant is changed
       - '"ans_test_create" in cm_add_tenant.proposed'
+      - cm_add_tenant.proposed.fvTenant.attributes.annotation == "orchestrator:ansible"
```
```suggestion
      - '"orchestrator:ansible" in cm_add_tenant.proposed'
```

sajagana reviewed (changes requested):
Also update the `tests/integration/targets/aci_rest/tasks/error_handling.yml` task: `- name: Verify error_on_input_validation`

```
-    - "error_on_input_validation.msg == 'APIC Error 801: property descr of tn-ansible_test failed validation for value \\'This is an [invalid] description\\''"
+    - "error_on_input_validation.msg == 'APIC Error 801: property descr of uni/tn-ansible_test failed validation for value \\'This is an [invalid] description\\''"
```

samiib commented:
> Also update the `tests/integration/targets/aci_rest/tasks/error_handling.yml` task: `- name: Verify error_on_input_validation`
> 
> ```
> -    - "error_on_input_validation.msg == 'APIC Error 801: property descr of tn-ansible_test failed validation for value \\'This is an [invalid] description\\''"
> +    - "error_on_input_validation.msg == 'APIC Error 801: property descr of uni/tn-ansible_test failed validation for value \\'This is an [invalid] description\\''"
> ```

@sajagana, Not sure about this one because the test was passing when I ran it myself. Just wondering which APIC did you use to test on?

akinross commented:
> > Also update the `tests/integration/targets/aci_rest/tasks/error_handling.yml` task: `- name: Verify error_on_input_validation`
> > ```
> > -    - "error_on_input_validation.msg == 'APIC Error 801: property descr of tn-ansible_test failed validation for value \\'This is an [invalid] description\\''"
> > +    - "error_on_input_validation.msg == 'APIC Error 801: property descr of uni/tn-ansible_test failed validation for value \\'This is an [invalid] description\\''"
> > ```
> 
> @sajagana, Not sure about this one because the test was passing when I ran it myself. Just wondering which APIC did you use to test on?

I'm seeing the same error now when running tests ( did not see runs before ). On november 28th the APICs where upgraded as I recall, perhaps the return message has changed?

samiib commented:
> > > Also update the `tests/integration/targets/aci_rest/tasks/error_handling.yml` task: `- name: Verify error_on_input_validation`
> > > ```
> > > -    - "error_on_input_validation.msg == 'APIC Error 801: property descr of tn-ansible_test failed validation for value \\'This is an [invalid] description\\''"
> > > +    - "error_on_input_validation.msg == 'APIC Error 801: property descr of uni/tn-ansible_test failed validation for value \\'This is an [invalid] description\\''"
> > > ```
> > 
> > 
> > @sajagana, Not sure about this one because the test was passing when I ran it myself. Just wondering which APIC did you use to test on?
> 
> I'm seeing the same error now when running tests ( did not see runs before ). On november 28th the APICs where upgraded as I recall, perhaps the return message has changed?

@sajagana & @akinross 

I have updated my branch with the latest changes from master and have been able to reproduce the failed assertion.
It appears the error message has changed between different versions of APIC. Some versions include `uni/` in the error and other version don't.
To account for this difference I changed the assertion to use a simple regex so the test will pass for this message variation.

sajagana reviewed (approved):
LGTM!

akinross reviewed (approved):
LGTM
